File: BinaryNotesMQ/java/src/org/bn/mq/impl/MessageQueue.java
/*
 * Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
 * 
 * Licensed under the LGPL, Version 2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.gnu.org/copyleft/lgpl.html
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * With any your questions welcome to my e-mail 
 * or blog at http://abdulla-a.blogspot.com.
 */

package org.bn.mq.impl;

import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import org.bn.mq.ICallAsyncListener;
import org.bn.mq.IConsumer;
import org.bn.mq.IPersistenceQueueStorage;
import org.bn.mq.IRemoteConsumer;
import org.bn.mq.IMessage;
import org.bn.mq.IMessageQueue;
import org.bn.mq.IQueue;
import org.bn.mq.net.ITransport;
import org.bn.mq.net.ITransportConnectionListener;
import org.bn.mq.net.ITransportReader;
import org.bn.mq.net.tcp.TransportPacket;
import org.bn.mq.protocol.DeliveredStatus;
import org.bn.mq.protocol.MessageBody;
import org.bn.mq.protocol.MessageEnvelope;
import org.bn.mq.protocol.SubscribeRequest;
import org.bn.mq.protocol.SubscribeResult;
import org.bn.mq.protocol.SubscribeResultCode;
import org.bn.mq.protocol.UnsubscribeResult;
import org.bn.mq.protocol.UnsubscribeResultCode;

public class MessageQueue<T> implements IMessageQueue<T>, Runnable, ITransportConnectionListener, ITransportReader {
    private ITransport transport;
    private String queuePath;
    private IQueue<T> queue = new Queue<T>();
    private AtomicInteger callCurId = new AtomicInteger(0);
    private Thread senderThread = new Thread(this);
    private AtomicBoolean stop = new AtomicBoolean(false);
    protected final Lock awaitMessageLock = new ReentrantLock();
    protected final Condition awaitMessageEvent  = awaitMessageLock.newCondition();
    protected Map<String,IConsumer<T> > consumers = new HashMap<String,IConsumer<T> >();
    protected Class<T> messageClass;
    protected IPersistenceQueueStorage<T> persistStorage;
    protected static AtomicInteger messageIdGenerator = new AtomicInteger(1);
        
    public MessageQueue(String queuePath, ITransport transport, Class<T> messageClass) {
        this.transport = transport;
        this.queuePath = queuePath;
        this.transport.addConnectionListener(this);
        this.transport.addReader(this);
        this.messageClass = messageClass;
        senderThread.setName("BNMessageQueue-"+queuePath);
        start();
    }
    
    public IQueue<T> getQueue() {
        return queue;
    }

    public void setQueue(IQueue<T> queue) {
        synchronized(queue) {
            this.queue = queue;
        }
    }

    public void sendMessage(IMessage<T> message) throws Exception {        
        if(message.getBody()==null)
            throw new Exception("Incorrect empty message body is specified to send!");
        awaitMessageLock.lock();
        synchronized(queue) {
            if(message.isMandatory())
                persistStorage.registerPersistenceMessage(message);
            queue.push(message);
        }        
        awaitMessageEvent.signal();
        awaitMessageLock.unlock();
    }
    
    public T call(T args, String consumerId, int timeout) throws Exception {
        Message<T> envelope =  new Message<T>(this.messageClass);
        envelope.setId(queuePath +"/call-"+callCurId.getAndIncrement());
        envelope.setBody(args);
        MessageEnvelope argsEnv = envelope.createEnvelope();
        MessageEnvelope result = null;
        argsEnv.getBody().getMessageUserBody().setQueuePath(this.getQueuePath());
        argsEnv.getBody().getMessageUserBody().setConsumerId(consumerId);
        IConsumer<T> consumer = null;
        synchronized(consumers) {
            if(!consumers.containsKey(consumerId)) {
                throw new Exception("Consumer with id:"+consumerId+" not found!");
            }
            else {
                 consumer = consumers.get(consumerId);
            }
        }
        if(consumer instanceof IRemoteConsumer) {
            ITransport consTransport = ((IRemoteConsumer<T>)consumer).getNetworkTransport();
            result = consTransport.call(argsEnv,timeout);
            envelope.fillFromEnvelope(result);
            return envelope.getBody();
        }
        else
            throw new Exception("Call enabled only for remote consumer !");
    }    

    public T call(T args, String consumerId) throws Exception {
        return call(args, consumerId, 120);
    }
    
    public void callAsync(T args, String consumerId, ICallAsyncListener<T> listener, int timeout) throws Exception {
        Message<T> envelope =  new Message<T>(this.messageClass);
        envelope.setId(queuePath +"/call-"+callCurId.getAndIncrement());
        envelope.setBody(args);
        MessageEnvelope argsEnv = envelope.createEnvelope();
        argsEnv.getBody().getMessageUserBody().setQueuePath(this.getQueuePath());
        argsEnv.getBody().getMessageUserBody().setConsumerId(consumerId);
        IConsumer<T> consumer = null;
        synchronized(consumers) {
            if(!consumers.containsKey(consumerId)) {
                throw new Exception("Consumer with id:"+consumerId+" not found!");
            }
            else {
                 consumer = consumers.get(consumerId);
            }
        }
        if(consumer instanceof IRemoteConsumer) {
            ITransport consTransport = ((IRemoteConsumer<T>)consumer).getNetworkTransport();
            consTransport.callAsync(argsEnv, new ProxyCallAsyncListener<T>(listener,messageClass),timeout );
        }
        else
            throw new Exception("Call enabled only for remote consumer !");
    }    

    public void callAsync(T args, String consumerId, ICallAsyncListener<T> listener) throws Exception {
        callAsync(args, consumerId, listener,120);
    }

    public void setPersistenseStorage(IPersistenceQueueStorage<T> storage) {
        this.persistStorage = storage;
    }

    public IPersistenceQueueStorage<T> getPersistenceStorage() {
        return this.persistStorage;
    }

    public void addConsumer(IConsumer<T> consumer) throws Exception {
        addConsumer(consumer, false);
    }

    public void addConsumer(IConsumer<T> consumer, Boolean persistence) throws Exception {
        addConsumer(consumer, persistence, null);
    }

    public void addConsumer(IConsumer<T> consumer, Boolean persistence, String filter) throws Exception {
        synchronized(consumers) {
            if(consumers.containsKey(consumer.getId())) {
                throw new Exception("Consumer with id:"+consumer.getId()+" is already subscription!");
            }
            else {
                consumers.put(consumer.getId(),consumer);
                if(persistence)
                    persistStorage.persistenceSubscribe(consumer);
            }
        }
    }

    public void delConsumer(IConsumer<T> consumer) throws Exception {
        synchronized(consumers) {
            if(!consumers.containsKey(consumer.getId())) {
                throw new Exception("Consumer with id:"+consumer.getId()+" doesn't have any subscription!");
            }
            else {
                consumers.remove(consumer.getId());
                persistStorage.persistenceUnsubscribe(consumer);                
            }        
        }
    }

    public String getQueuePath() {
        return queuePath;
    }

    public void run() {        
        IMessage<T> message = null;
        do {
            awaitMessageLock.lock();
            synchronized(queue) {
                message = queue.getNext();
            }
            if(message==null) {
                
                try {
                    awaitMessageEvent.await();
                }
                catch(Exception ex) {ex =null; }                
            }            
            awaitMessageLock.unlock();
            if(message!=null) {
                synchronized(consumers) {
                    for(Map.Entry<String, IConsumer<T> > entry: consumers.entrySet()) {
                        entry.getValue().onMessage(message);
                    }
                }
            }
            
        }
        while(!stop.get());
    }

    public synchronized void stop() {
        stop.set(true);        
        try {
            if(senderThread.isAlive()) {
                awaitMessageLock.lock();
                awaitMessageEvent.signal();
                awaitMessageLock.unlock();
                this.transport.delConnectionListener(this);
                this.transport.delReader(this);
                senderThread.join();
            }
        }
        catch (InterruptedException e) {
             e.printStackTrace();
        }
    }

    public synchronized void start() {
        stop.set(false);
        if(!senderThread.isAlive()) {
            senderThread.start();
        }
        
    }
    
    protected void dispose() {        
        stop();
    }
    
    private void onReceiveSubscribeRequest(MessageEnvelope message, ITransport transport) {
        RemoteConsumer<T> remoteConsumer = new RemoteConsumer<T>(message.getBody().getSubscribeRequest().getConsumerId(), transport, this.messageClass);
        
        MessageEnvelope resultMsg = new MessageEnvelope();
        MessageBody body = new MessageBody();
        SubscribeResult subscribeResult = new SubscribeResult();
        SubscribeResultCode subscribeResultCode = new SubscribeResultCode();
        body.selectSubscribeResult(subscribeResult);
        resultMsg.setBody(body);
        resultMsg.setId(message.getId());            
        try {                
            addConsumer(remoteConsumer,message.getBody().getSubscribeRequest().getPersistence(),message.getBody().getSubscribeRequest().getFilter());
            subscribeResultCode.setValue(SubscribeResultCode.EnumType.success);
        }
        catch (Exception e) {
            subscribeResultCode.setValue(SubscribeResultCode.EnumType.alreadySubscription);
            subscribeResult.setDetails(e.toString());
        }
        subscribeResult.setCode(subscribeResultCode);
        try {
            transport.sendAsync(resultMsg);
            if(subscribeResultCode.getValue() == SubscribeResultCode.EnumType.success 
            && message.getBody().getSubscribeRequest().getPersistence()) {
                List<IMessage<T>> messages =  persistStorage.getMessagesToSend(remoteConsumer);
                awaitMessageLock.lock();
                synchronized(queue) {
                    queue.push(messages);
                }        
                awaitMessageEvent.signal();
                awaitMessageLock.unlock();
            }
            
        }
        catch (Exception e) {
            e.printStackTrace();
        }    
    }

    private void onReceiveUnsubscribeRequest(MessageEnvelope message, ITransport transport) {
        IConsumer<T> consumer = null;
        synchronized(consumers) {
            consumer = consumers.get(message.getBody().getUnsubscribeRequest().getConsumerId());
        }
        MessageEnvelope resultMsg = new MessageEnvelope();
        MessageBody body = new MessageBody();
        UnsubscribeResult unsubscribeResult = new UnsubscribeResult();
        UnsubscribeResultCode unsubscribeResultCode = new UnsubscribeResultCode();
        body.selectUnsubscribeResult(unsubscribeResult);
        resultMsg.setBody(body);
        resultMsg.setId(message.getId());            
        
        if(consumer!=null) {            
            try {
                delConsumer(consumer);
            }
            catch (Exception e) {
                // TODO
            }
            unsubscribeResultCode.setValue(UnsubscribeResultCode.EnumType.success);
        }
        else {
            unsubscribeResultCode.setValue(UnsubscribeResultCode.EnumType.subscriptionNotExists);
        }
        unsubscribeResult.setCode(unsubscribeResultCode);
        try {
            transport.sendAsync(resultMsg);
        }
        catch (Exception e) {
            e.printStackTrace();
        }    
    }
    
    private void onReceivedDeliveryReport(MessageEnvelope message, ITransport transport) {
        try {
            if(message.getBody().getDeliveryReport().getStatus().getValue() == DeliveredStatus.EnumType.delivered ) {
                persistStorage.removeDeliveredMessage(
                    message.getBody().getDeliveryReport().getConsumerId(),
                    message.getBody().getDeliveryReport().getMessageId()
                );
            }            
        }
        catch (Exception e) {
            e.printStackTrace();
        }        
    }
    

    public boolean onReceive(MessageEnvelope message, ITransport transport) {
        if(message.getBody().isSubscribeRequestSelected() && message.getBody().getSubscribeRequest().getQueuePath().equalsIgnoreCase(queuePath) ) {
            onReceiveSubscribeRequest(message,transport);
            return true;
        }
        else
        if(message.getBody().isUnsubscribeRequestSelected() && message.getBody().getUnsubscribeRequest().getQueuePath().equalsIgnoreCase(queuePath) ) {
            onReceiveUnsubscribeRequest(message,transport);
            return true;
        }
        else
        if(message.getBody().isDeliveryReportSelected() && message.getBody().getDeliveryReport().getQueuePath().equalsIgnoreCase(queuePath) ) {
            onReceivedDeliveryReport(message,transport);
        }
        return false;
    }

    public void onConnected(ITransport transport) {
    }

    public void onDisconnected(ITransport transport) {
        synchronized(consumers) {
            Map<String,IConsumer<T> > newConsumers = new HashMap<String,IConsumer<T> >();
            for(Map.Entry<String,IConsumer<T>  > entry : consumers.entrySet()) {
                if(entry.getValue() instanceof IRemoteConsumer ) {
                    IRemoteConsumer<T> consumer = (IRemoteConsumer<T>) entry.getValue();
                    if(!consumer.getNetworkTransport().equals(transport)) {
                        newConsumers.put(entry.getValue().getId(),entry.getValue());
                    }                
                }
            }
            consumers.clear();
            consumers.putAll(newConsumers);
        }
    }
    
    public IMessage<T> createMessage(T body) {
        Message<T> result = new Message<T>(this.messageClass);
        result.setBody(body);
        result.setId(this.getQueuePath()+"/#"+new Date().getTime()+"/"+messageIdGenerator.getAndIncrement());
        result.setQueuePath(getQueuePath());
        return result;
    }
    
    public IMessage<T> createMessage() {
        return createMessage(null);
    }    
}

File: BinaryNotesMQ/.net/BinaryNotesMQ/src/org/bn/mq/impl/MessageQueue.cs
/*
 Copyright 2006-2011 Abdulla Abdurakhmanov (abdulla@latestbit.com)
 Original sources are available at www.latestbit.com

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
using System;
using System.Threading;
using System.Collections.Generic;
using org.bn.mq.protocol;
using org.bn.mq.net;

namespace org.bn.mq.impl
{
	
	public class MessageQueue<T> : IMessageQueue<T>, ITransportConnectionListener, ITransportReader
	{
        private IQueue<T> queue = new Queue<T>();
		virtual public IQueue<T> Queue
		{
			get
			{
				return queue;
			}
			
			set
			{
				lock (queue)
				{
					this.queue = value;
				}
			}			
		}
        private static long messageIdGenerator = 1;

        protected internal IPersistenceQueueStorage<T> persistStorage;
		virtual public IPersistenceQueueStorage<T> PersistenceStorage
		{
			set
			{
				this.persistStorage = value;
			}

            get
            {
                return this.persistStorage;
            }
			
		}

        virtual public System.String QueuePath
		{
			get
			{
				return queuePath;
			}
			
		}
		private ITransport transport;
		private System.String queuePath;
		
		private int callCurId = 0;
		private Thread senderThread;

		private bool stopThreadFlag = true;
		protected internal AutoResetEvent awaitMessageEvent = new AutoResetEvent(false);
		protected internal IDictionary<String, IConsumer<T> > consumers = new Dictionary<String, IConsumer<T>>();
		
		public MessageQueue(string queuePath, ITransport transport)
		{
            senderThread = new Thread(new ThreadStart(this.Run));
            this.transport = transport;
			this.queuePath = queuePath;
			this.transport.addConnectionListener(this);
			this.transport.addReader(this);
			senderThread.Name = "BNMessageQueue-" + queuePath;
			start();
		}
		
		public virtual void  sendMessage(IMessage<T> message)
		{
			if (message.Body == null)
				throw new Exception("Incorrect empty message body is specified to send!");
			lock (queue)
			{
				if (message.Mandatory)
					persistStorage.registerPersistenceMessage(message);
				queue.push(message);
			}
			awaitMessageEvent.Set();
		}
		public virtual T call(T args, string consumerId, int timeout)
		{
            Message<T> envelope = new Message<T>();
            envelope.Id = (queuePath + "/call-" + callCurId++);
            envelope.Body = (args);
            MessageEnvelope argsEnv = envelope.createEnvelope();
            argsEnv.Body.MessageUserBody.QueuePath = this.QueuePath;
            argsEnv.Body.MessageUserBody.ConsumerId = consumerId;
            
            IConsumer<T> consumer = null;
			lock (consumers)
			{
				if (!consumers.ContainsKey(consumerId))
				{
					throw new Exception("Consumer with id:" + consumerId + " not found!");
				}
				else
				{
					consumer = consumers[consumerId];
				}
			}
			
            if (consumer is IRemoteConsumer<T>)
			{
                MessageEnvelope result = null;
				ITransport consTransport = ((IRemoteConsumer<T>)consumer).NetworkTransport;
				result = consTransport.call(argsEnv, timeout);
				envelope.fillFromEnvelope(result);
				return envelope.Body;
			}
			else
				throw new System.Exception("Call enabled only for remote consumer !");
		}
		
		public virtual T call(T args, string consumerId)
		{
			return call(args, consumerId, 120);
		}

        public virtual void callAsync(T args, string consumerId, ICallAsyncListener<T> listener, int timeout)
		{
            Message<T> envelope = new Message<T>();
            envelope.Id = (queuePath + "/call-" + callCurId++);
            envelope.Body = (args);
            MessageEnvelope argsEnv = envelope.createEnvelope();
            argsEnv.Body.MessageUserBody.QueuePath = this.QueuePath;
            argsEnv.Body.MessageUserBody.ConsumerId = consumerId;
            IConsumer<T> consumer = null;
            lock (consumers)
			{
				if (!consumers.ContainsKey(consumerId))
				{
					throw new Exception("Consumer with id:" + consumerId + " not found!");
				}
				else
				{
					consumer = consumers[consumerId];
				}
			}
            if (consumer is IRemoteConsumer<T>)
			{
				ITransport consTransport = ((IRemoteConsumer<T>)consumer).NetworkTransport;
				consTransport.callAsync(argsEnv, new ProxyCallAsyncListener<T>( listener), timeout);
			}
			else
				throw new System.Exception("Call enabled only for remote consumer !");
		}


        public virtual void callAsync(T args, string consumerId, ICallAsyncListener<T> listener)
		{
			callAsync(args, consumerId, listener, 120);
		}

        public virtual void callAsync(T args, string consumerId, CallAsyncDelegateResult<T> resultDelegate, CallAsyncDelegateTimeout<T> timeoutDelegate)
        {
            callAsync(args, consumerId, resultDelegate, timeoutDelegate, 120);
        }

        public virtual void callAsync(T args, string consumerId, CallAsyncDelegateResult<T> resultDelegate, CallAsyncDelegateTimeout<T> timeoutDelegate, int timeout)
        {
            Message<T> envelope = new Message<T>();
            envelope.Id = (queuePath + "/call-" + callCurId++);
            envelope.Body = (args);
            MessageEnvelope argsEnv = envelope.createEnvelope();
            argsEnv.Body.MessageUserBody.QueuePath = this.QueuePath;
            argsEnv.Body.MessageUserBody.ConsumerId = consumerId;
            IConsumer<T> consumer = null;
            lock (consumers)
            {
                if (!consumers.ContainsKey(consumerId))
                {
                    throw new Exception("Consumer with id:" + consumerId + " not found!");
                }
                else
                {
                    consumer = consumers[consumerId];
                }
            }
            if (consumer is IRemoteConsumer<T>)
            {
                ITransport consTransport = ((IRemoteConsumer<T>)consumer).NetworkTransport;
                consTransport.callAsync(argsEnv, new ProxyCallAsyncListener<T>(resultDelegate, timeoutDelegate), timeout);
            }
            else
                throw new System.Exception("Call enabled only for remote consumer !");
        }

		
		public virtual void  addConsumer(IConsumer<T> consumer)
		{
			addConsumer(consumer, false);
		}
		
		public virtual void  addConsumer(IConsumer<T> consumer, bool persistence)
		{
			addConsumer(consumer, persistence, null);
		}
		
		public virtual void  addConsumer(IConsumer<T> consumer, bool persistence, string filter)
		{
			lock (consumers)
			{
				if (consumers.ContainsKey(consumer.Id))
				{
					throw new Exception("Consumer with id:" + consumer.Id + " is already subscription!");
				}
				else
				{
					consumers[consumer.Id] = consumer;
					if (persistence)
						persistStorage.persistenceSubscribe(consumer);
				}
			}
		}
		
		public virtual void  delConsumer(IConsumer<T> consumer)
		{
			lock (consumers)
			{
                persistStorage.persistenceUnsubscribe(consumer);
				if (!consumers.ContainsKey(consumer.Id))
				{
					throw new Exception("Consumer with id:" + consumer.Id + " doesn't have any subscription!");
				}
				else
				{
					consumers.Remove(consumer.Id);
				}
			}
		}
		
		public virtual void  Run()
		{
			IMessage<T> message = null;
			do 
			{
				lock (queue)
				{
					message = queue.getNext();
				}
				if (message == null)
				{					
					awaitMessageEvent.WaitOne();
				}
                else
				{
					lock (consumers)
					{
						foreach(IConsumer<T> entry in consumers.Values)
						{
							entry.onMessage(message);
						}
					}
				}
			}
			while (!stopThreadFlag);
		}
		
		public virtual void  stop()
		{
			lock (this)
			{				
				try
				{
                    if (!stopThreadFlag)//if (senderThread.IsAlive) - don't support CF
					{
                        stopThreadFlag = true;
						awaitMessageEvent.Set();
						this.transport.delConnectionListener(this);
						this.transport.delReader(this);
						senderThread.Join();
					}
				}
				catch (Exception e)
				{
					Console.WriteLine(e);
				}
			}
		}
		
		public virtual void  start()
		{
			lock (this)
			{
				
				if (stopThreadFlag)//!senderThread.IsAlive) CF not support isAlive
				{
                    stopThreadFlag = false;
					senderThread.Start();
				}
			}
		}

        ~MessageQueue() 
        {
            stop();
        }
		
		private void  onReceiveSubscribeRequest(MessageEnvelope message, ITransport transport)
		{
			RemoteConsumer<T> remoteConsumer = new RemoteConsumer<T>(
                message.Body.SubscribeRequest.ConsumerId, transport
            );
			
			MessageEnvelope resultMsg = new MessageEnvelope();
			MessageBody body = new MessageBody();
			SubscribeResult subscribeResult = new SubscribeResult();
			SubscribeResultCode subscribeResultCode = new SubscribeResultCode();
			body.selectSubscribeResult(subscribeResult);
			resultMsg.Body = body;
			resultMsg.Id = message.Id;
			try
			{
				addConsumer(remoteConsumer, message.Body.SubscribeRequest.Persistence, message.Body.SubscribeRequest.Filter);
				subscribeResultCode.Value = SubscribeResultCode.EnumType.success;
			}
			catch (System.Exception e)
			{
				subscribeResultCode.Value = SubscribeResultCode.EnumType.alreadySubscription;
				subscribeResult.Details = (e.ToString());
			}
			subscribeResult.Code = subscribeResultCode;
			try
			{
				transport.sendAsync(resultMsg);

                if (subscribeResult.Code.Value == SubscribeResultCode.EnumType.success 
                    && message.Body.SubscribeRequest.Persistence)
                {
                    IList<IMessage<T>> messages = persistStorage.getMessagesToSend(remoteConsumer);
                    lock (queue)
                    {
                        queue.push(messages);
                    }
                    awaitMessageEvent.Set();
                }

			}
			catch (System.Exception e)
			{
				Console.WriteLine(e);
			}
		}
		
		private void  onReceiveUnsubscribeRequest(MessageEnvelope message, ITransport transport)
		{
			IConsumer<T> consumer = null;
			lock (consumers)
			{
                if(consumers.ContainsKey(message.Body.UnsubscribeRequest.ConsumerId))
				    consumer = consumers[message.Body.UnsubscribeRequest.ConsumerId];
			}

			MessageEnvelope resultMsg = new MessageEnvelope();
			MessageBody body = new MessageBody();
			UnsubscribeResult unsubscribeResult = new UnsubscribeResult();
			UnsubscribeResultCode unsubscribeResultCode = new UnsubscribeResultCode();
			body.selectUnsubscribeResult(unsubscribeResult);
			resultMsg.Body = body;
			resultMsg.Id = message.Id;
			
			if (consumer != null)
			{
				try
				{
					delConsumer(consumer);
				}
				catch (System.Exception e)
				{
					// TODO
				}
				unsubscribeResultCode.Value = UnsubscribeResultCode.EnumType.success;
			}
			else
			{
				unsubscribeResultCode.Value = UnsubscribeResultCode.EnumType.subscriptionNotExists;
			}
			unsubscribeResult.Code = unsubscribeResultCode;
			try
			{
				transport.sendAsync(resultMsg);
			}
			catch (System.Exception e)
			{
				Console.WriteLine(e);
			}
		}

        private void onReceiveDeliveryReport(MessageEnvelope message, ITransport transport)
        {
            try
            {
                if (message.Body.DeliveryReport.Status.Value == DeliveredStatus.EnumType.delivered)
                {
                    persistStorage.removeDeliveredMessage(
                        message.Body.DeliveryReport.ConsumerId,
                        message.Body.DeliveryReport.MessageId
                    );
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }        
        }

		public virtual bool onReceive(MessageEnvelope message, ITransport transport)
		{
			if (message.Body.isSubscribeRequestSelected() && message.Body.SubscribeRequest.QueuePath.ToUpper().Equals(queuePath.ToUpper()))
			{
				onReceiveSubscribeRequest(message, transport);
				return true;
			}
            else
			if (message.Body.isUnsubscribeRequestSelected() && message.Body.UnsubscribeRequest.QueuePath.ToUpper().Equals(queuePath.ToUpper()))
			{
				onReceiveUnsubscribeRequest(message, transport);
				return true;
			}
            else
            if (message.Body.isDeliveryReportSelected() && message.Body.DeliveryReport.QueuePath.ToUpper().Equals(queuePath.ToUpper()))
            {
                onReceiveDeliveryReport(message, transport);
                return true;
            }
			return false;
		}
		
		public virtual void  onConnected(ITransport transport)
		{
		}
		
		public virtual void  onDisconnected(ITransport transport)
		{
			lock (consumers)
			{
                IDictionary<String, IConsumer<T> > newConsumers = new Dictionary<String, IConsumer<T>>();
				foreach(IConsumer<T> entry in consumers.Values)
				{
					if (entry is IRemoteConsumer<T>)
					{
						IRemoteConsumer<T> consumer = (IRemoteConsumer<T>)entry;
						if (!consumer.NetworkTransport.Equals(transport))
						{
							newConsumers[entry.Id] = entry;
						}
					}
				}
				consumers.Clear();
				putAll(newConsumers, consumers);
			}
		}

        private void putAll(IDictionary<string, IConsumer<T>> source, IDictionary<string, IConsumer<T>> target)
        {
            IEnumerator<KeyValuePair<string, IConsumer<T>>> enumerator = source.GetEnumerator();
            while (enumerator.MoveNext())
            {
                KeyValuePair<string, IConsumer<T>> keyPair = enumerator.Current;
                target.Add(keyPair.Key, keyPair.Value);
            }
        }
		
		public virtual IMessage<T> createMessage(T body)
		{
			Message<T> result = new Message<T>();
			result.Body = body;
			result.Id = this.QueuePath+"/#"+(DateTime.Now.Ticks/10000).ToString()+"/"+messageIdGenerator++;
			result.QueuePath = this.QueuePath;
			return result;
		}
		
		public virtual IMessage<T> createMessage()
		{
			return createMessage(default(T));
		}
	}
}
